src/backend/distributed/commands/function.c

```diff
@@ -113,6 +119,13 @@ create_distributed_function(PG_FUNCTION_ARGS)
 	UpdateFunctionDistributionInfo(&functionAddress, distributionArgumentIndex,
 								   colocationId);
 
+	/*
+	 * Once we have at least one distributed function/procedure, we sync the
+	 * metadata to nodes so that the function/procedure delegation can be handled
+	 * locally on the nodes.
+	 */
+	TriggerSyncMetadataToPrimaryNodes();
```

Comment:
1. We already do `SendCommandToWorkers()` few lines above, so doing metadata sync here (and not postponing it to maintenanced) won't add any transactional limitations.
2. In addition to that, this is wasteful, since it causes **all** metadata on workers to be dropped and recreated from scratch.

Reply:
>We already do SendCommandToWorkers() few lines above, so doing metadata sync here (and not delegating it to maintenanced) won't add any transactional limitations.

We don't support:
```SQL
BEGIN;
SELECT start_metadata_Sync_to_node('localhost', 9700);
ERROR:  start_metadata_sync_to_node cannot run inside a transaction block
```

So, if we sync metadata within `create_distributed_function()/master_add_node()`, we'd impose these limitations on those UDFs as well. That's why, doing it lazily seems a better idea.

>n addition to that, this is wasteful, since it causes all metadata on workers to be dropped and recreated from scratch.

Is it? The code looks like we only to `if (workerNode->hasMetadata && !workerNode->metadataSynced)` [here](https://github.com/citusdata/citus/blob/ff100b2720e3efed2241b969ca560c86859851a6/src/backend/distributed/metadata/metadata_sync.c#L1334). Am I missing anything?

Note: `TriggerSyncMetadataToPrimaryNodes` only syncs when `if (!workerNode->hasMetadata)`, so maybe I should rename the function to reflect it?

Reply:
It's possible that the transaction restriction is there only for historical reasons. It might be ok to sync the metadata snapshot as part of a distributed transaction. That said, I'm fine with either approach.

Reply:
I don't want to remove that check now because we'd need to think/test a bit more. For example, we should switch to sequential mode. Or, what happens if `create_distributed_function` is followed by `SELECT` call in the same transaction? For now, we seem fine what we have. 


Now, we're just not optimizing some `SELECT/CALL` pushdowns, so seems safer.